cocoa/src/toga_cocoa/widgets/numberinput.py

```diff
@@ -158,65 +169,83 @@ def create(self):
         self.add_constraints()
 
     def set_color(self, color):
-        self.input.textColor = native_color(color)
+        self.native_input.textColor = native_color(color)
 
-    def set_readonly(self, value):
-        # Even if it's not editable, it's still selectable.
-        self.input.editable = not value
-        self.input.selectable = True
+    def set_background_color(self, color):
+        if color is TRANSPARENT:
+            # The text view needs to be made transparent *and* non-bezeled
+            self.native_input.drawsBackground = False
+            self.native_input.bezeled = False
+        else:
+            self.native_input.drawsBackground = True
+            self.native_input.bezeled = True
+            self.native_input.backgroundColor = native_color(color)
+
+    def has_focus(self):
+        # When the NSTextField gets focus, a field editor is created, and that editor
+        # has the original widget as the delegate. The first responder is the Field Editor.
+        return isinstance(self.native.window.firstResponder, NSTextView) and (
+            self.native.window.firstResponder.delegate == self.native_input
+        )
+
+    def focus(self):
+        if not self.has_focus():
+            self.interface.window._impl.native.makeFirstResponder(self.native_input)
 
-    def set_placeholder(self, value):
-        self.input.cell.placeholderString = value
+    def get_readonly(self):
+        return not self.native_input.isEditable()
+
+    def set_readonly(self, value):
+        self.native_input.editable = not value
 
     def set_step(self, step):
-        self.stepper.increment = self.interface.step
+        self.native_stepper.increment = step
 
     def set_min_value(self, value):
-        if self.interface.min_value is None:
-            self.stepper.minValue = -sys.maxsize
+        if value is None:
+            self.native_stepper.minValue = -sys.float_info.max
         else:
-            self.stepper.minValue = value
+            self.native_stepper.minValue = float(value)
 
     def set_max_value(self, value):
-        if self.interface.max_value is None:
-            self.stepper.maxValue = sys.maxsize
+        if value is None:
+            self.native_stepper.maxValue = sys.float_info.max
         else:
-            self.stepper.maxValue = value
+            self.native_stepper.maxValue = float(value)
 
     def set_alignment(self, value):
-        self.input.alignment = NSTextAlignment(value)
+        self.native_input.alignment = NSTextAlignment(value)
 
     def set_font(self, font):
-        if font:
-            self.input.font = font._impl.native
+        self.native_input.font = font._impl.native
+
+    def get_value(self):
+        try:
+            raw = str(self.native_input.stringValue)
+            return Decimal(raw).quantize(self.interface.step)
```

Comment:
Decimal uses `ROUND_HALF_EVEN` mode by default, e.g. 1.25 rounds to 1.2, and 1.35 rounds to 1.4. This probably isn't what the user expects.

Also, this pattern will be repeated in each backend, so it should be factored out.